src/clients/c/tb_client.zig

```diff
@@ -49,6 +53,77 @@ const TestingContext = blk: {
     break :blk ContextType(EchoClient);
 };
 
+/// Logging is global per application; it would be nice to be able to define a different logger for
+/// each client instance, though.
+var logging: Logging = .{};
+
+pub const Logging = struct {
+    const Callback = *const fn (
+        message_level: u8,
+        message_ptr: [*]const u8,
+        message_len: usize,
+    ) callconv(.C) void;
+
+    const log_line_max = 8192;
+
+    /// A logger which defers to an application provided handler.
+    pub fn application_logger(
+        comptime message_level: std.log.Level,
+        comptime scope: @Type(.EnumLiteral),
+        comptime format: []const u8,
+        args: anytype,
+    ) void {
+        // Messages are silently dropped if no logging callback is specified - unless they're warn
+        // or err. The value in having those for debugging is too high to silence them, even until
+        // client libraries catch up and implement a callback handler.
+        const callback = logging.callback orelse {
+            if (message_level == .warn or message_level == .err) {
+                std.log.defaultLog(message_level, scope, format, args);
+            }
+            return;
+        };
+
+        // Protect everything with a mutex - logging can be called from different threads
+        // simultaneously, and there's only one buffer for now.
+        logging.mutex.lock();
+        defer logging.mutex.unlock();
+
+        const prefix = if (scope == .default) ": " else "(" ++ @tagName(scope) ++ "): ";
+        const output = std.fmt.bufPrint(
+            &logging.buffer,
+            prefix ++ format,
+            args,
+        ) catch |err| switch (err) {
+            error.NoSpaceLeft => blk: {
+                // Print an error indicating the log message has been truncated, before the
+                // truncated log itself.
+                const message = "the following log message has been truncated:";
+                callback(@intFromEnum(std.log.Level.err), message.ptr, message.len);
+
+                break :blk &logging.buffer;
+            },
+            else => unreachable,
+        };
+
+        callback(@intFromEnum(message_level), output.ptr, output.len);
+    }
+
+    callback: ?Callback = null,
+    mutex: std.Thread.Mutex = .{},
+    buffer: [log_line_max]u8 = undefined,
+};
+
+pub fn register_log_callback(
```

Comment:
There's no way to unregister or register a new handler, currently.

Reply:
We could make the API simpler by allowing the user to replace the handler, or even setting it as `null` if they want to unregister.